### PR TITLE
Switch Python SDK harness entrypoint to use python3

### PR DESCRIPTION
**Pull request by Prabin120**
Beam's Python SDK harness boot entrypoint may fail when `python` is not accessible, but `python3` is.

fixes: #26958

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Prabin120** [comment]
Successfully changes is being made by changing python to python3. Beam's python SDK issue #27236 solved

**Comment by tvalentyn** [comment]
Run Dataflow Python ValidatesContainer

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesContainer

**Comment by tvalentyn** [comment]
Thanks a lot for your contribution, @Prabin120. We can merge once the testsuite passes.

**Comment by tvalentyn** [comment]
Actually, after some thinking, I think `python` is a valid command, and is referenced in official documents: https://docs.python.org/3/using/cmdline.html#command-line. 

@Prabin120 I would suggest creating a helper function that identifies the python interpreter dynamically and falls back to to `python3` if `python` is not available. We could run that function once per and reuse the return string in other calls.

**Comment by tvalentyn** [comment]
we would be able to further use the helper function to make the interpreter configurable via an environment variable.

**Review by Prabin120 — COMMENTED** [review]
As your suggestion I have created a helper function which will help to give the write python version if available

File: sdks/python/container/piputil.go [hunk 1]
```diff
@@ -28,7 +28,19 @@ import (
 
 	"github.com/apache/beam/sdks/v2/go/pkg/beam/util/execx"
 )
-
+// It will return the python interpreter type i.e. python or python3 if available else return empty string.
+func pythonVersion() string {
```
Comment (tvalentyn) [inline sdks/python/container/piputil.go:32]: Let's not duplicate the implementation. Can we find a home for the helper under https://github.com/apache/beam/tree/master/sdks/go/pkg/beam/util/ ?

Comment (tvalentyn) [reply sdks/python/container/piputil.go:32]: Actually, looks like there is one already: https://github.com/apache/beam/blob/1446295d66d394114d4f3e99980a598b9608881c/sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go#L385

**Review by Prabin120 — COMMENTED** [review]
Changes is being made and have used the pre defined function to check the python version

**Comment by Prabin120** [comment]
Thank you sir for pointing ou the pre defined function to check python version. I have made changes according to the function. Please verify.

File: sdks/python/container/piputil.go [hunk 2]
```diff
@@ -50,8 +38,8 @@ func pipInstallRequirements(files []string, dir, name string) error {
 			// option will make sure that only things staged in the worker will be
 			// used without following their dependencies.
 			args := []string{"-m", "pip", "install", "-r", filepath.Join(dir, name), "--no-cache-dir", "--disable-pip-version-check", "--no-index", "--no-deps", "--find-links", dir}
-			pythonVersion := pythonVersion()
-			if pythonVersion != "" {
+			pythonVersion,err := expansionx.getPythonVersion()
+			if err == nil {
```
Comment (tvalentyn) [inline sdks/python/container/piputil.go:42]: we need to fail with a clear error otherwise.

File: sdks/python/container/boot.go [hunk 3]
```diff
@@ -330,7 +334,7 @@ func setupVenv(ctx context.Context, logger *tools.Logger, baseDir, workerId stri
 	if err := os.MkdirAll(dir, 0750); err != nil {
 		return "", fmt.Errorf("failed to create Python venv directory: %s", err)
 	}
-	if err := execx.Execute("python", "-m", "venv", "--system-site-packages", dir); err != nil {
+	if err := execx.Execute("python3", "-m", "venv", "--system-site-packages", dir); err != nil {
```
Comment (tvalentyn) [inline sdks/python/container/boot.go:337]: need to use helper here as well.

File: sdks/python/container/boot.go [hunk 4]
```diff
@@ -91,11 +92,14 @@ func main() {
 			"--container_executable=/opt/apache/beam/boot",
 		}
 		log.Printf("Starting worker pool %v: python %v", workerPoolId, strings.Join(args, " "))
-		if err := execx.Execute("python", args...); err != nil {
-			log.Fatalf("Python SDK worker pool exited with error: %v", err)
+		pythonVersion,err := expansionx.getPythonVersion()
+		if err == nil{
```
Comment (tvalentyn) [inline sdks/python/container/boot.go:96]: I think a more common pattern is to handle error case, then proceed with normal flow:

```
if pythonVersion, err := expansionx.getPythonVersion(); err != nil {
    return err //  (or some other message + err)
}
<non-error flow>
```
any concerns with that ? thank you!

File: sdks/python/container/piputil.go [hunk 5]
```diff
@@ -78,17 +82,17 @@ func pipInstallPackage(files []string, dir, name string, force, optional bool, e
 				// installed if necessary.  This achieves our goal outlined above.
 				args := []string{"-m", "pip", "install", "--no-cache-dir", "--disable-pip-version-check", "--upgrade", "--force-reinstall", "--no-deps",
 					filepath.Join(dir, packageSpec)}
-				err := execx.Execute("python", args...)
+				err := execx.Execute("python3", args...)
```
Comment (tvalentyn) [inline sdks/python/container/piputil.go:85]: use the helper here as well and all other applicable places.

**Review by Prabin120 — COMMENTED** [review]
All the changes is being made

**Comment by Prabin120** [comment]
All the changes is being made successfully.

File: sdks/python/container/boot.go [hunk 6]
```diff
@@ -91,12 +92,11 @@ func main() {
 			"--container_executable=/opt/apache/beam/boot",
 		}
 		log.Printf("Starting worker pool %v: python %v", workerPoolId, strings.Join(args, " "))
-		if err := execx.Execute("python", args...); err != nil {
-			log.Fatalf("Python SDK worker pool exited with error: %v", err)
+		if pythonVersion, err := expansionx.getPythonVersion(); err != nil {
+			log.Print("Python SDK worker pool exited.")
+			os.Exit(0)
 		}
```
Comment (damccorm) [inline sdks/python/container/boot.go:98]: We should be calling `execx.Execute(pythonVersion, args...)` here, right?

**Comment by tvalentyn** [comment]
waiting on author

**Comment by tvalentyn** [comment]
Hi @Prabin120 , what are next steps on this PR? Do you need any help?

**Comment by tvalentyn** [comment]
closing in favor of https://github.com/apache/beam/pull/28046

**Pull request closed without merging**